### ISSUE TITLE
[Bug] Gemma4 prefill materializes a quadratic dense image-attention mask

### ISSUE THREAD

**Issue by wangshang23** [sgl-project/sglang#33033]
## Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists on the tested upstream `main` revision.
- [x] Environment information and a reproducible benchmark configuration are
  included.
- [x] This report is in English.

## Describe the bug

Gemma4 multimodal prefill currently materializes a dense attention mask with
shape `[extend_seq_len, extend_seq_len + prefix_len]` to make the soft tokens
of each image bidirectional. For a request without a prefix, this is an `S x S`
mask even though the non-causal region needed for one image is only a small
image-token block (256 tokens in this test).

The mask is initialized, triangularized, flattened, concatenated, stored as a
generic `custom_mask`, and then read tile-by-tile by the Triton extend-attention
kernel. Its metadata and memory traffic therefore grow as `O(S^2)` rather than
with the number of images.

The same generic mask is also supplied to every decoder attention layer. Gemma4
only requires bidirectional image-token attention in sliding-attention layers;
full-attention layers should remain causal, consistent with the direction of
PR `#32303`.

PR `#32303` addresses the layer-scoping semantics. This report focuses on the
remaining quadratic dense representation and Triton custom-mask path.

On H20, replacing the dense mask with compact image spans leaves text prefill
effectively unchanged and reduces vision TTFT increasingly as sequence length
grows:

| Total tokens | Vision before | Vision after | Reduction | Speedup |
| ---: | ---: | ---: | ---: | ---: |
| 8k | 1105.432 ms | 936.549 ms | 15.3% | 1.18x |
| 16k | 3030.869 ms | 2327.110 ms | 23.2% | 1.30x |
| 32k | 9483.857 ms | 6721.119 ms | 29.1% | 1.41x |

After subtracting paired text TTFT, the removed vision overhead is 170.403 ms at
8k, 704.551 ms at 16k, and 2824.313 ms at 32k. It grows by 4.13x and 4.01x as
sequence length doubles, matching the expected quadratic dense-mask scaling.

A second controlled A/B on NVIDIA DGX Spark shows the same length-dependent
effect. Both sides use upstream `5c6635d8f`; the patched side adds only commit
`0ebaa3f2e`. The model uses FP16 vision and W4A16 compressed-tensors language
weights.

| Total tokens | Vision before | Vision after | Reduction | Speedup |
| ---: | ---: | ---: | ---: | ---: |
| 1000 | 313.058 ms | 287.546 ms | 8.15% | 1.09x |
| 1664 | 544.840 ms | 482.214 ms | 11.49% | 1.13x |
| 4000 | 1556.841 ms | 1067.809 ms | 31.41% | 1.46x |
| 8000 | 4201.442 ms | 2245.855 ms | 46.55% | 1.87x |

The DGX requests used one `336 x 336` image with exactly 256 image tokens and
the reported length is the exact combined image-plus-text input length. The
dominant-color output remained `orange`/`Orange` before and after the change.
Decode latency stayed in the same 22.8--24.6 ms/token range and is not part of
the performance claim.

## Reproduction

Model and request setup:

- Gemma4 E4B BF16
- Triton attention backend
- One `336 x 336` square image, producing 256 image tokens
- Total input lengths of 8000, 16000, and 32000 tokens
- Text and vision requests alternated in one engine
- One warmup and three measured requests for each case
- TTFT measured from `Engine.generate(..., stream=True)` start to the first chunk
- 16 output tokens, `temperature=0`, `ignore_eos=True`

Relevant engine configuration:

```python
Engine(
    model_path="/path/to/gemma-4-E4B-it",
    trust_remote_code=True,
    enable_multimodal=True,
    device="cuda",
    dtype="bfloat16",
    attention_backend="triton",
    sampling_backend="pytorch",
    grammar_backend="none",
    max_total_tokens=36864,
    context_length=34816,
    max_prefill_tokens=32768,
    max_running_requests=1,
    chunked_prefill_size=32768,
    disable_overlap_schedule=True,
    disable_radix_cache=True,
    swa_full_tokens_ratio=1.0,
    disable_cuda_graph=True,
    skip_server_warmup=True,
    mem_fraction_static=0.7,
)
```

All three requests run as a single complete prefill; scheduler logs report
exactly 8k, 16k, or 32k new tokens in one batch.

The DGX Spark A/B uses the same streaming timing method with exact total input
lengths of 1000, 1664, 4000, and 8000 tokens. It uses Triton attention, PyTorch
sampling, one warmup, three measured requests, 16 output tokens, and disables
the VLM cache, overlap scheduling, and radix cache. Decode CUDA graph is
enabled for batch size 1; prefill CUDA graph is disabled. The effective
`chunked_prefill_size` is 8192, so all four requests execute as one complete
prefill.

The source of the quadratic allocation is
`python/sglang/srt/models/gemma4_mm.py::prepare_attn_masks`, which creates the
dense tensor before exposing it through `forward_metadata.custom_mask`.

Expected behavior: pass compact per-request image intervals and evaluate
`causal or same_image` directly in the Triton kernel. Metadata should be
`O(num_images)`, and image spans should only be enabled for sliding-attention
layers.

## Environment

```text
SGLang revision: 425349b79 (upstream main at test time)
Python: 3.12.3
GPU: NVIDIA H20, compute capability 9.0
CUDA toolkit: 13.1
CUDA driver: 535.161.08
PyTorch: 2.11.0+cu130
sglang: 0.5.9
sglang-kernel: 0.4.5
Triton: 3.5.1
Transformers: 5.12.1
FlashInfer: 0.6.4
```

Additional controlled DGX Spark A/B:

```text
SGLang base revision: 5c6635d8f
Patched revision: 5c6635d8f + 0ebaa3f2e
GPU: NVIDIA GB10 (DGX Spark)
PyTorch: 2.11.0+cu130
sglang-kernel: 0.4.5
Transformers: 5.10.1
Model precision: FP16 vision + W4A16 compressed-tensors language model
```

**Comment by efschu** [comment]
Supporting data from the other end of the same path. Before optimising the dense mask, note that a large fraction
of Gemma4 prefills install a mask that is **bit-for-bit the causal predicate the Triton kernel already applies**:
a multi-turn follow-up whose image span sits wholly in the cached prefix, a chunked-prefill chunk that splits a
span, and a single-token span all produce a `tril` with no bidirectional bit written. Tracking whether any such bit
was written and skipping the install is free and removes the buffer entirely for those batches. One trap: the skip
has to be at **batch** granularity — the mask is one flat buffer addressed by `mask_indptr` and `USE_CUSTOM_MASK`
is a whole-launch `constexpr`, so a per-request skip leaves a zero-length slot and the kernel reads the next
request's bytes with this request's row stride. Also worth knowing when benchmarking: sglang's own VLM boot warmup
posts a 32×32 base64 PNG, so the "warmup prefill" is a 275-token *image* request (256 soft tokens), not a text one.
Analysis and tests in https://github.com/efschu/htsglang (commit `09b37bfbbc`).

**Comment by efschu** [comment]
Opened https://github.com/sgl-project/sglang/pull/33261 — the correctness/no-op half only, not the quadratic allocation you are asking about. It skips installing the mask when no bidirectional bit was written for any request in the batch, which is bit-for-bit the causal predicate the extend kernel already applies on its `IS_CAUSAL` branch. One constraint worth knowing if you optimize the buffer itself: the skip is only safe at batch granularity, because the mask is one flat buffer addressed by `mask_indptr` and `USE_CUSTOM_MASK` is a whole-launch constexpr, so a per-request skip leaves a zero-length slot and the kernel reads the next request's bytes at this request's row stride. The buffer is still allocated and discarded in the skipped case, so your issue stays open on its own terms.
